tensorboard/plugins/pr_curve/pr_curve_demo.py

```diff
@@ -0,0 +1,225 @@
+  examples = tf.concat([true_reds, true_greens, true_blues], axis=0)
+  probabilities_colors_are_red = (1 - red_predictor.cdf(
+      tf.norm(examples - tf.constant([255., 0, 0]), axis=1))) * 2
+  probabilities_colors_are_green = (1 - green_predictor.cdf(
+      tf.norm(examples - tf.constant([0, 255., 0]), axis=1))) * 2
+  probabilities_colors_are_blue = (1 - blue_predictor.cdf(
+      tf.norm(examples - tf.constant([0, 0, 255.]), axis=1))) * 2
+
+  predictions = (
+      probabilities_colors_are_red,
+      probabilities_colors_are_green,
+      probabilities_colors_are_blue
+  )
+
+  # This is the crucial piece. We write data required for generating PR curves.
+  # We create 1 summary per class because we create 1 PR curve per class.
+  for i, color in enumerate(('red', 'green', 'blue')):
+    description = ('The probabilities used to create this PR curve are '
+                   'generated from a normal distribution. Its standard '
+                   'deviation is initially %0.0f and decreases over time.' %
+                       initial_standard_deviations[i])
+
+    weights = None
+    if assign_random_weights:
+      weights = tf.random_uniform(shape=tf.shape(predictions[i]), maxval=1.0)
+
+    summary.op(
+        tag=color,
+        labels=labels[:, i],
+        predictions=predictions[i],
+        num_thresholds=thresholds,
+        weights=weights,
+        display_name='classifying %s' % color,
+        description=description)
+  merged_summary_op = tf.summary.merge_all()
+  events_directory = os.path.join(logdir, run_name)
+  sess = tf.Session()
+  writer = tf.summary.FileWriter(events_directory, sess.graph)
+
+  for step in xrange(steps):
+    feed_dict = {
+        iteration: step,
+    }
+    merged_summary = sess.run(merged_summary_op, feed_dict=feed_dict)
+    writer.add_summary(merged_summary, step)
+
+  writer.close()
+
+
+def run_all(logdir, steps, thresholds, verbose=False):
+  """Generate PR curve summaries.
+
+  Arguments:
+    logdir: The directory into which to store all the runs' data.
+    steps: The number of steps to run for.
+    verbose: Whether to print the names of runs into stdout during execution.
+    thresholds: The number of thresholds to use for PR curves.
+  """
+  # First, we generate data for a PR curve that assigns even weights for
+  # predictions of all classes.
+  run_name = 'colors'
+  if verbose:
+      print('--- Running: %s' % run_name)
+  start_runs(
+      logdir=logdir,
+      steps=steps,
+      run_name=run_name,
+      thresholds=thresholds)
+
+  # Next, we generate data for a PR curve that assigns 2x the weight for
+  # predictions of whether a color is blue.
+  run_name = 'colors_weighted_randomly'
+  if verbose:
+      print('--- Running: %s' % run_name)
+  start_runs(
+      logdir=logdir,
+      steps=steps,
+      run_name=run_name,
+      thresholds=thresholds,
+      assign_random_weights=True)
+
+
+def main(unused_argv):
+  logdir = os.path.expanduser(FLAGS.logdir)
```

Comment:
FWIW, I strongly disagree with the use of `expanduser` in the TensorBoard binary, as well. Applications don't expand paths, full-stop. It should not be common at all to pass `--logdir "~/Desktop"`—quoted—and in the case where a user does that we should respect the user's choice and use the folder `./~/Desktop`. (Otherwise, there's no way for them to specify that folder!)